### ISSUE TITLE
SessionStore tries to fetch a member before a DB connection is established

### ISSUE THREAD

**Issue by indygriffiths** [silverstripe/silverstripe-mfa#274]
Framework: 3.7.3
PHP: 7.1
Module version: 3.0.x-dev (378401b3185e403706916cc8f7c6e89ed3c96a8a)

When attempting to log in after installing this module and the TOTP authenticator the login process fails on the `Security/LoginForm/mfa` endpoint.

From the looks of it the MFA SessionStore is attempting to grab the Member object before the DB connection has been established, which is causing the `DB::get_conn()` call inside `Convert.php` to fail.

<details>
<summary>Error</summary>

```
Warning: session_start(): Failed to decode session object. Session has been destroyed in /var/www/html/platform-dashboard/framework/control/Session.php on line 382

Fatal error: Uncaught Error: Call to a member function quoteString() on null in /var/www/html/platform-dashboard/framework/core/Convert.php:171 
Stack trace: 
#0 /var/www/html/platform-dashboard/framework/model/DataQuery.php(285): Convert::raw2sql('Member', true) 
#1 /var/www/html/platform-dashboard/framework/model/DataQuery.php(479): DataQuery->getFinalisedQuery() 
#2 /var/www/html/platform-dashboard/framework/model/DataList.php(819): DataQuery->firstRow() 
#3 /var/www/html/platform-dashboard/framework/model/DataObject.php(3287): DataList->first() 
#4 /var/www/html/platform-dashboard/framework/model/DataObject.php(3377): DataObject::get_one('Member', Array, true) 
#5 /var/www/html/platform-dashboard/mfa/src/Store/SessionStore.php(215): DataObject::get_by_id('Member', 277) 
#6 [internal function]: SilverStripe\MFA\Store\SessionStore->unserialize('{"member":277,"...') 
#7 /var/www/html/platform-dashboard/framework/control/Session.php(382): session_start() 
#8 /var/www/html/platform-dashboard/framework/control/Session.php(598): Sessi in /var/www/html/platform-dashboard/framework/core/Convert.php on line 171
```
</details>

<details>
<summary>Backtrace from `Convert.php`</summary>

```
array(22) {
  [0]=>
  array(5) {
    ["file"]=>
    string(62) "/var/www/html/platform-dashboard/framework/model/DataQuery.php"
    ["line"]=>
    int(285)
    ["function"]=>
    string(7) "raw2sql"
    ["class"]=>
    string(7) "Convert"
    ["type"]=>
    string(2) "::"
  }
  [1]=>
  array(5) {
    ["file"]=>
    string(62) "/var/www/html/platform-dashboard/framework/model/DataQuery.php"
    ["line"]=>
    int(479)
    ["function"]=>
    string(17) "getFinalisedQuery"
    ["class"]=>
    string(9) "DataQuery"
    ["type"]=>
    string(2) "->"
  }
  [2]=>
  array(5) {
    ["file"]=>
    string(61) "/var/www/html/platform-dashboard/framework/model/DataList.php"
    ["line"]=>
    int(819)
    ["function"]=>
    string(8) "firstRow"
    ["class"]=>
    string(9) "DataQuery"
    ["type"]=>
    string(2) "->"
  }
  [3]=>
  array(5) {
    ["file"]=>
    string(63) "/var/www/html/platform-dashboard/framework/model/DataObject.php"
    ["line"]=>
    int(3287)
    ["function"]=>
    string(5) "first"
    ["class"]=>
    string(8) "DataList"
    ["type"]=>
    string(2) "->"
  }
  [4]=>
  array(5) {
    ["file"]=>
    string(63) "/var/www/html/platform-dashboard/framework/model/DataObject.php"
    ["line"]=>
    int(3377)
    ["function"]=>
    string(7) "get_one"
    ["class"]=>
    string(10) "DataObject"
    ["type"]=>
    string(2) "::"
  }
  [5]=>
  array(5) {
    ["file"]=>
    string(63) "/var/www/html/platform-dashboard/mfa/src/Store/SessionStore.php"
    ["line"]=>
    int(214)
    ["function"]=>
    string(9) "get_by_id"
    ["class"]=>
    string(10) "DataObject"
    ["type"]=>
    string(2) "::"
  }
  [6]=>
  array(3) {
    ["function"]=>
    string(11) "unserialize"
    ["class"]=>
    string(35) "SilverStripe\MFA\Store\SessionStore"
    ["type"]=>
    string(2) "->"
  }
  [7]=>
  array(3) {
    ["file"]=>
    string(62) "/var/www/html/platform-dashboard/framework/control/Session.php"
    ["line"]=>
    int(382)
    ["function"]=>
    string(13) "session_start"
  }
  [8]=>
  array(5) {
    ["file"]=>
    string(62) "/var/www/html/platform-dashboard/framework/control/Session.php"
    ["line"]=>
    int(598)
    ["function"]=>
    string(10) "inst_start"
    ["class"]=>
    string(7) "Session"
    ["type"]=>
    string(2) "->"
  }
  [9]=>
  array(5) {
    ["file"]=>
    string(76) "/var/www/html/platform-dashboard/testsession/code/TestSessionEnvironment.php"
    ["line"]=>
    int(62)
    ["function"]=>
    string(5) "start"
    ["class"]=>
    string(7) "Session"
    ["type"]=>
    string(2) "::"
  }
  [10]=>
  array(3) {
    ["function"]=>
    string(11) "__construct"
    ["class"]=>
    string(22) "TestSessionEnvironment"
    ["type"]=>
    string(2) "->"
  }
  [11]=>
  array(5) {
    ["file"]=>
    string(80) "/var/www/html/platform-dashboard/framework/control/injector/InjectionCreator.php"
    ["line"]=>
    int(20)
    ["function"]=>
    string(11) "newInstance"
    ["class"]=>
    string(15) "ReflectionClass"
    ["type"]=>
    string(2) "->"
  }
  [12]=>
  array(5) {
    ["file"]=>
    string(72) "/var/www/html/platform-dashboard/framework/control/injector/Injector.php"
    ["line"]=>
    int(560)
    ["function"]=>
    string(6) "create"
    ["class"]=>
    string(16) "InjectionCreator"
    ["type"]=>
    string(2) "->"
  }
  [13]=>
  array(5) {
    ["file"]=>
    string(72) "/var/www/html/platform-dashboard/framework/control/injector/Injector.php"
    ["line"]=>
    int(866)
    ["function"]=>
    string(11) "instantiate"
    ["class"]=>
    string(8) "Injector"
    ["type"]=>
    string(2) "->"
  }
  [14]=>
  array(5) {
    ["file"]=>
    string(56) "/var/www/html/platform-dashboard/testsession/_config.php"
    ["line"]=>
    int(4)
    ["function"]=>
    string(3) "get"
    ["class"]=>
    string(8) "Injector"
    ["type"]=>
    string(2) "->"
  }
  [15]=>
  array(4) {
    ["file"]=>
    string(75) "/var/www/html/platform-dashboard/framework/core/manifest/ConfigManifest.php"
    ["line"]=>
    int(144)
    ["args"]=>
    array(1) {
      [0]=>
      string(56) "/var/www/html/platform-dashboard/testsession/_config.php"
    }
    ["function"]=>
    string(12) "require_once"
  }
  [16]=>
  array(5) {
    ["file"]=>
    string(58) "/var/www/html/platform-dashboard/framework/core/Config.php"
    ["line"]=>
    int(324)
    ["function"]=>
    string(14) "activateConfig"
    ["class"]=>
    string(17) "SS_ConfigManifest"
    ["type"]=>
    string(2) "->"
  }
  [17]=>
  array(5) {
    ["file"]=>
    string(56) "/var/www/html/platform-dashboard/framework/core/Core.php"
    ["line"]=>
    int(129)
    ["function"]=>
    string(22) "pushConfigYamlManifest"
    ["class"]=>
    string(6) "Config"
    ["type"]=>
    string(2) "->"
  }
  [18]=>
  array(4) {
    ["file"]=>
    string(51) "/var/www/html/platform-dashboard/framework/main.php"
    ["line"]=>
    int(151)
    ["args"]=>
    array(1) {
      [0]=>
      string(56) "/var/www/html/platform-dashboard/framework/core/Core.php"
    }
    ["function"]=>
    string(12) "require_once"
  }
  [19]=>
  array(3) {
    ["file"]=>
    string(77) "/var/www/html/platform-dashboard/framework/core/startup/ErrorControlChain.php"
    ["line"]=>
    int(215)
    ["function"]=>
    string(9) "{closure}"
  }
  [20]=>
  array(5) {
    ["file"]=>
    string(77) "/var/www/html/platform-dashboard/framework/core/startup/ErrorControlChain.php"
    ["line"]=>
    int(206)
    ["function"]=>
    string(4) "step"
    ["class"]=>
    string(17) "ErrorControlChain"
    ["type"]=>
    string(2) "->"
  }
  [21]=>
  array(5) {
    ["file"]=>
    string(51) "/var/www/html/platform-dashboard/framework/main.php"
    ["line"]=>
    int(188)
    ["function"]=>
    string(7) "execute"
    ["class"]=>
    string(17) "ErrorControlChain"
    ["type"]=>
    string(2) "->"
  }
}

```
</details>

**Comment by robbieaverill** [comment]
Triaged as medium impact because I think this only affects users with the testsession module installed (which is designed for local development).

**Comment by ScopeyNZ** [comment]
Oh I prepped a patch for this locally that I haven't pushed up yet. I'll push it up to the 4 branch and it can be combined with backporting later on.

**Comment by robbieaverill** [comment]
Merged #277 into 4.0, will be backported soon
